### PR TITLE
[#415] Configure collection view cells

### LINKED ISSUE #415: Create a collection view cell for products
- [ ] Collection view should be passed its collection to set title
- [ ] Create a horizontal collection view which will hold the products within that collection
- [ ] Show loading state and placeholders when downloading products
- [ ] Have collections view controller download the specific collection's products and use callback from data provider to pass its downloaded products (data provider -> VC -> collection view cell)

### PR DESCRIPTION
**Pull request by EishanVijay**
Collection view container cells are configured to display collection name and products within that specific collection.

A BaseOperation class was created to decouple Operation handling from the view model classes. Collection and product view models will subclass this BaseOperation class in order to handle product and collection operations.

Both view models extend `UICollectionViewDataSource` because protocol methods deal with model objects and we only want the view models to handle model objects.

Both collection and product cells are 'dumb' and only know how to configure themselves with data provided by the view models.

Product label animations were taken from Apple's UIKitCatalog tvOS sample app.

Resolves #415 

@davidmuzi @bgulanowski

### CODE REVIEW

File: Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/ProductCollectionViewCell.swift [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    func configure(title: String) {
+        self.productTitleLabel.text = title
+    }
+    
+    override func didUpdateFocus(in context: UIFocusUpdateContext, with coordinator: UIFocusAnimationCoordinator) {
+        /*
+         Update the label's alpha value using the `UIFocusAnimationCoordinator`.
+         This will ensure all animations run alongside each other when the focus
+         changes.
+         */
```
Comment (davidmuzi) [inline Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/ProductCollectionViewCell.swift]: comment does not add much value, Id suggest removing

File: Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/ProductCollectionViewCell.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        coordinator.addCoordinatedAnimations({
+            if self.isFocused {
```
Comment (davidmuzi) [inline Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/ProductCollectionViewCell.swift]: Can be expressed using a ternary operator self.producTitleLabel.alpha = (self.focussed) ? 1.0 : 0.0

File: Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/BaseOperation.swift [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+//  BaseOperation.swift
+//  Mobile Buy SDK tvOS Sample App
+//
+//  Created by Eishan Vijay on 2016-11-23.
+//  Copyright © 2016 Shopify. All rights reserved.
+//
+
+import Foundation
+
+class BaseOperation: NSObject {
```
Comment (davidmuzi) [inline Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/BaseOperation.swift]: BaseViewModel?

File: Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/CollectionsViewModel.swift [hunk 4]
```diff
@@ -27,29 +27,40 @@
...
+        }
+        self.setCurrentOperation(operation: operation)
+    }
+}
+
+extension CollectionsViewModel: UICollectionViewDataSource {
+    
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: CollectionViewContainerCell.reuseIdentifier, for: indexPath) as? CollectionViewContainerCell
+        let collection = self.collections[indexPath.section] as BUYCollection
```
Comment (davidmuzi) [inline Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/CollectionsViewModel.swift]: the type cast is not needed, `collections` is already declared as `[Collection]`

File: Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/CollectionsViewModel.swift [hunk 5]
```diff
@@ -27,29 +27,40 @@
...
+        let productsViewModel = ProductsViewModel(collection: collection, dataProvider: self.dataProvider)
+        cell?.configure(viewModel: productsViewModel, title: collection.title)
```
Comment (davidmuzi) [inline Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/CollectionsViewModel.swift]: No need to be defensive here, the cell should be implicitly unwrapped.

`let cell = collectionView.dequeueReusableCell(withReuseIdentifier: CollectionViewContainerCell.reuseIdentifier, for: indexPath) as! CollectionViewContainerCell`

File: Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/ProductsViewModel.swift [hunk 6]
```diff
@@ -42,18 +41,30 @@ class Products {
...
-            self.collectionView.reloadData()
-        }
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return self.products.count
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: ProductCollectionViewCell.reuseIdentifier, for: indexPath) as? ProductCollectionViewCell
+        let product = self.products[indexPath.row] as BUYProduct
+        cell?.configure(title: product.title)
```
Comment (davidmuzi) [inline Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/ProductsViewModel.swift]: same as in collections

File: Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/ProductCollectionViewCell.swift (hunk 1 above)
Comment (EishanVijay) [inline Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/ProductCollectionViewCell.swift]: It's removed in another PR

File: Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/CollectionsViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    }
+}
+
+extension CollectionsViewModel: UICollectionViewDataSource {
+    
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: CollectionViewContainerCell.reuseIdentifier, for: indexPath) as! CollectionViewContainerCell
+        let collection = self.collections[indexPath.section]
+        let productsViewModel = ProductsViewModel(collection: collection, dataProvider: self.dataProvider)
+        cell.configure(viewModel: productsViewModel, title: collection.title)
```
Comment (davidmuzi) [inline Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/CollectionsViewModel.swift]: why does the configure method need the title parameter if the viewModel already has the `collection`?

Comment (EishanVijay) [inline Mobile Buy SDK Sample Apps/Mobile Buy SDK tvOS Sample/Mobile Buy SDK tvOS Sample/CollectionsViewModel.swift]: Giving the cell its title here removes the need to write a getter on ProductsViewModel to get the collection's title

**Pull request merged**